### PR TITLE
Add simple ByteTokenizer for Reformer

### LINKED ISSUE #11649: Reformer inference widget is broken
From: https://huggingface.co/google/reformer-enwik8?text=My+name+is+Julien+and+I+like+to

image

@patrickvonplaten

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes #11649

Adds a ReformerByteTokenizer for (https://huggingface.co/google/reformer-enwik8).

- Everything is a bit hardcoded in that tokenizer as very little configuration is expected
- no fast tokenizer (a bit useless as this manipulates raw bytes
  and has very little python logic)

Added to the docs
Added very simple tests. For this tokenizer, the actual mixin is debattable as
"tokens" are raw bytes, and cannot be strings (255 for instance is not a valid string).
using b"\xff" instead of 255 is possible yet might not be exactly clearer.

This requires some modifications within the "google/reformer-enwik8" config.
Namely:
 - Adding a `tokenizer_class` to the config
 - Adding a dummy file so that AutoTokenizer won't fail because no file are needed.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik  @patrickvonplaten

### CODE REVIEW

File: tests/test_tokenization_reformer.py [hunk 1]
```diff
@@ -328,3 +328,65 @@ def test_torch_encode_plus_sent_to_model(self):
         with torch.no_grad():
             model(**encoded_sequence)
             model(**batch_encoded_sequence)
+
+
+class ReformerByteTokenizationTest(unittest.TestCase):
```
Comment (LysandreJik) [inline tests/test_tokenization_reformer.py:333]: I would also at least test saving and loading, the `__call__` method, that it can be sent to a model, padding, truncation, ...

File: src/transformers/models/reformer/tokenization_reformer.py [hunk 2]
```diff
@@ -167,3 +167,35 @@ def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] =
             copyfile(self.vocab_file, out_vocab_file)
 
         return (out_vocab_file,)
+
+
+class ReformerByteTokenizer(PreTrainedTokenizer):
```
Comment (LysandreJik) [inline src/transformers/models/reformer/tokenization_reformer.py:172]: Is that byte tokenizer free of special tokens?

Comment (LysandreJik) [reply src/transformers/models/reformer/tokenization_reformer.py:172]: The tokenizer is added to the docs, but there isn't really any doc. What is that tokenizer? How can it be used? With what should it be used?

The current full docs looks like this:

image

Comment (LysandreJik) [reply src/transformers/models/reformer/tokenization_reformer.py:172]: If it doesn't need a vocabulary, then it would be nice to indicate this somewhere too - and probably to remove the requirement for the tokenizer to even have a vocabulary file in the first place. This would make saving/loading simpler and wouldn't rely on dummy files.

If it's too hard because it would mean overriding the `from_pretrained` method, then we can keep the workaround documented.

**Review by LysandreJik — COMMENTED** [review]
Cool, thank you for opening this PR! 

Before merging, I think two issues should be resolved:
- I'm not sure what that `ReformerByteTokenizer` can be used for - so if I'm unaware, I suppose users will be too. Adding a bit of documentation regarding what is that tokenizer and why it's here would be nice
- According to what it can be used for, the appropriate tests should be put in place. For example there's no testing for saving/loading while there's a bit of a workaround to enable that - it would be nice to test that and all the other expected behaviors.

**Comment by Narsil** [comment]
> Cool, thank you for opening this PR!
> 
> Before merging, I think two issues should be resolved:
> 
>     * I'm not sure what that `ReformerByteTokenizer` can be used for - so if I'm unaware, I suppose users will be too. Adding a bit of documentation regarding what is that tokenizer and why it's here would be nice

Tried to add some documentation, is it better ? Ultimately I don't really have background for this model, so I added more or less what was said in the model card (and left a link to it just in case, I'm also unaware of other models that are Byte level too.).

> 
>     * According to what it can be used for, the appropriate tests should be put in place. For example there's no testing for saving/loading while there's a bit of a workaround to enable that - it would be nice to test that and all the other expected behaviors.

I added it. Because it has only 1 special token (pad), which I arbitrarily set to 0 (does not seem to be explained in the model card either currently, but why would there be a 2 shift for input_ids otherwise ?) and it does not really make sense for this Tokenizer to try to use "tokens" in the current Tokenizer meaning (substrings that are part of the main string). It is impossible to do because of how utf-8 works, If we did we would run sooner or later into other issues:

`len(chr(255).encode("utf-8")) == 2` for instance which is `chr(255) == [195, 195]`, not `[255]`.

**Comment by LysandreJik** [comment]
@patrickvonplaten can you give this one a look and merge if it looks good to you?
